### PR TITLE
[LSP] Report unchanged results if diagnostic data is the same as last report

### PR DESCRIPTION
**Pull request by dibarbet**
Resolves https://github.com/dotnet/roslyn/issues/75240

This updates the behavior of the version cache to also store and check against the hash of the last reported data.  This allows us to report unchanged results when the data we re-calculate ends up being exactly the same.

This saves some overhead in serialization / deserialization and processing on the client side.

### CODE REVIEW

File: src/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs [hunk 1]
```diff
@@ -156,18 +158,18 @@ protected virtual Task WaitForChangesAsync(string? category, RequestContext cont
...
-                    computeCheapVersionAsync: async () => (globalStateVersion, await project.GetDependentVersionAsync(cancellationToken).ConfigureAwait(false)),
-                    computeExpensiveVersionAsync: async () => (globalStateVersion, await project.GetDependentChecksumAsync(cancellationToken).ConfigureAwait(false)),
+                    cacheState,
                     cancellationToken).ConfigureAwait(false);
-                if (newResultId != null)
+                if (newResult != null)
                 {
-                    await ComputeAndReportCurrentDiagnosticsAsync(
-                        context, diagnosticSource, progress, newResultId, clientCapabilities, cancellationToken).ConfigureAwait(false);
+                    ReportCurrentDiagnostics(
```
Comment (dibarbet) [inline src/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs:171]: Computation moved to run inside the cache based on what parameters it determines

File: src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    /// <summary>
+    /// Cache where we store the data produced by prior requests so that they can be returned if nothing of significance 
+    /// changed. The <see cref="VersionStamp"/> is produced by <see cref="Project.GetDependentVersionAsync(CancellationToken)"/> while the 
+    /// <see cref="Checksum"/> is produced by <see cref="Project.GetDependentChecksumAsync(CancellationToken)"/>.  The former is faster
+    /// and works well for us in the normal case.  The latter still allows us to reuse diagnostics when changes happen that
+    /// update the version stamp but not the content (for example, forking LSP text).
+    /// </summary>
+    internal class DiagnosticsPullCache(string uniqueKey) : VersionedPullCache<(int globalStateVersion, VersionStamp? dependentVersion), (int globalStateVersion, Checksum dependentChecksum), DiagnosticsRequestState, DiagnosticData>(uniqueKey)
+    {
+        public override async Task<(int globalStateVersion, VersionStamp? dependentVersion)> ComputeCheapVersionAsync(DiagnosticsRequestState state, CancellationToken cancellationToken)
```
Comment (dibarbet) [inline src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs:27]: moved from passing functions to impls of abstract methods

File: src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler;
+
+internal abstract partial class VersionedPullCache<TCheapVersion, TExpensiveVersion, TState, TComputedData>
+{
+    /// <summary>
+    /// Internal cache item that updates state for a particular <see cref="Workspace"/> and <see cref="ProjectOrDocumentId"/> in <see cref="VersionedPullCache{TCheapVersion, TExpensiveVersion, TState, TComputedData}"/>
+    /// This type ensures that the state for a particular key is never updated concurrently for the same key (but different key states can be concurrent).
```
Comment (dibarbet) [inline src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs:17]: prior to this change, there was a single semaphore around updating resultIds for the entire cache item.  However now we need to also run the computation to determine if and how the cache should be updated.  I did not want to block requests for other documents on computation of diagnostics for a different diagnostic.  So now we have the `CacheItem` which gives a finer grained semaphore to an entry for a specific workspace+proj/doc

**Comment by CyrusNajmabadi** [comment]
LMK when you'd like a review :)

File: src/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs [hunk 4]
```diff
@@ -290,12 +290,10 @@ private async Task ComputeAndReportCurrentDiagnosticsAsync(
             return;
         }
 
-        context.TraceInformation($"Found {diagnostics.Length} diagnostics for {diagnosticSource.ToDisplayString()}");
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs:293]: don't want this message?  or did this move?

File: src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Roslyn.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+
+internal abstract partial class AbstractPullDiagnosticHandler<TDiagnosticsParams, TReport, TReturn> where TDiagnosticsParams : IPartialResultParams<TReport>
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs:14]:
```suggestion
internal abstract partial class AbstractPullDiagnosticHandler<TDiagnosticsParams, TReport, TReturn>
    where TDiagnosticsParams : IPartialResultParams<TReport>
```

nit

File: src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    internal record struct DiagnosticsRequestState(Project Project, int GlobalStateVersion, RequestContext Context, IDiagnosticSource DiagnosticSource);
+
+    /// <summary>
+    /// Cache where we store the data produced by prior requests so that they can be returned if nothing of significance 
+    /// changed. The <see cref="VersionStamp"/> is produced by <see cref="Project.GetDependentVersionAsync(CancellationToken)"/> while the 
+    /// <see cref="Checksum"/> is produced by <see cref="Project.GetDependentChecksumAsync(CancellationToken)"/>.  The former is faster
+    /// and works well for us in the normal case.  The latter still allows us to reuse diagnostics when changes happen that
+    /// update the version stamp but not the content (for example, forking LSP text).
+    /// </summary>
+    internal class DiagnosticsPullCache(string uniqueKey) : VersionedPullCache<(int globalStateVersion, VersionStamp? dependentVersion), (int globalStateVersion, Checksum dependentChecksum), DiagnosticsRequestState, DiagnosticData>(uniqueKey)
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs:25]: sealed?  possibly private?

File: src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            return (state.GlobalStateVersion, await state.Project.GetDependentVersionAsync(cancellationToken).ConfigureAwait(false));
+        }
+
+        public override async Task<(int globalStateVersion, Checksum dependentChecksum)> ComputeExpensiveVersionAsync(DiagnosticsRequestState state, CancellationToken cancellationToken)
+        {
+            return (state.GlobalStateVersion, await state.Project.GetDependentChecksumAsync(cancellationToken).ConfigureAwait(false));
+        }
+
+        public override async Task<ImmutableArray<DiagnosticData>> ComputeDataAsync(DiagnosticsRequestState state, CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs:37]: possibly doc this or the base about what sort of locks/blockign is happening while this is called.

File: src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            var diagnostics = await state.DiagnosticSource.GetDiagnosticsAsync(state.Context, cancellationToken).ConfigureAwait(false);
+            state.Context.TraceInformation($"Found {diagnostics.Length} diagnostics for {state.DiagnosticSource.ToDisplayString()}");
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs:40]: ah ok.  it moved here.

File: src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        ///   unimportant changes happen that trigger the cheap version change detection.</item>
+        ///   <item>The hash of the data that was computed when the resultId was generated.
+        ///       <para>
+        ///       When the versions above change, we must recalculate the data.  However sometimes that data ends up being exactly the same as the prior request.
+        ///       When that happens, this allows us to send back an unchanged result instead of reserializing data the client already has.
+        ///       </para>
+        ///   </item>
+        /// </list>
+        /// 
+        /// </summary>
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs:46]: Nit: consider mentioning this is guarded by _gate.

File: src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        public async Task<(string, ImmutableArray<TComputedData>)?> UpdateCacheItemAsync(
+            VersionedPullCache<TCheapVersion, TExpensiveVersion, TState, TComputedData> cache,
+            PreviousPullResult? previousPullResult,
+            bool isFullyLoaded,
+            TState state,
+            CancellationToken cancellationToken)
+        {
+            // Ensure that we only update the cache item one at a time.
+            // This means that the computation of new data for this item only occurs sequentially.
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs:62]: aside: have we considered computing results in parallel *higher up* in workspace diagnostics?  we walk the diagnostic sources serially.  but we could use our parallel helpers to potentially speed that up.

File: src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+        /// <summary>
+        /// Updates the values for this cache entry.  Guarded by <see cref="_semaphore"/>
+        /// 
+        /// Returns null if the previousPullResult can be re-used, otherwise returns a new resultId and the new data associated with it.
+        /// </summary>
+        public async Task<(string, ImmutableArray<TComputedData>)?> UpdateCacheItemAsync(
+            VersionedPullCache<TCheapVersion, TExpensiveVersion, TState, TComputedData> cache,
+            PreviousPullResult? previousPullResult,
+            bool isFullyLoaded,
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs:57]: hrmm... what is isFullyLoaded...

File: src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        ///   </item>
+        /// </list>
+        /// 
+        /// </summary>
+        private (string resultId, TCheapVersion cheapVersion, TExpensiveVersion expensiveVersion, int hashedData)? _lastResult;
+
+        /// <summary>
+        /// Updates the values for this cache entry.  Guarded by <see cref="_semaphore"/>
+        /// 
+        /// Returns null if the previousPullResult can be re-used, otherwise returns a new resultId and the new data associated with it.
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs:52]: Use langword for null.

File: src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+                var data = await cache.ComputeDataAsync(state, cancellationToken).ConfigureAwait(false);
+                var dataHash = GetComputedDataHash(data);
+
+                string newResultId;
+                if (_lastResult is not null && _lastResult?.resultId == previousPullResult?.PreviousResultId && _lastResult?.hashedData == dataHash)
+                {
+                    // The new data we've computed is exactly the same as the data we computed last time even though the versions have changed.
+                    // Instead of reserializing everything, we can return the same result id back to the client.
+
+                    // Ensure we store the updated versions we calculated with the old resultId.
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs:109]: could you mention if this update is actually necessary or not?   wasn't clear to me if that was the case.

File: src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+                    newResultId = $"{uniqueKey}:{cache.IncrementResultId()}";
+                    _lastResult = (newResultId, cheapVersion, expensiveVersion, dataHash);
+                    return (newResultId, data);
+                }
+            }
+        }
+
+        private static int GetComputedDataHash(ImmutableArray<TComputedData> data)
+        {
+            var hashes = data.SelectAsArray(d => d.GetHashCode()).Sort();
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs:134]: needs docs.  first, why are you sorting?

second.  are these types HashCode safe?  Not convinced they are.  i have some thoughts here.  overall idea is sound, but if we don't make this part bulletproof we may not send updates when we need to.

File: src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.cs [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.cs:14]: fsn.

File: src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.cs [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+{
+    /// <summary>
+    /// Specialized cache used by the 'pull' LSP handlers.  Supports storing data to know when to tell a client
+    /// that existing results can be reused, or if new results need to be computed.  Multiple keys can be used,
+    /// with different computation costs to determine if the previous cached data is still valid.
+    /// </summary>
+    internal abstract partial class VersionedPullCache<TCheapVersion, TExpensiveVersion, TState, TComputedData>(string uniqueKey) where TComputedData : notnull
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.cs:21]: wrap constraints due to long line.

File: src/LanguageServer/Protocol/Handler/SpellCheck/SpellCheckPullCache.cs [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+
+    public override async Task<ImmutableArray<SpellCheckSpan>> ComputeDataAsync(SpellCheckState state, CancellationToken cancellationToken)
+    {
+        var spans = await state.Service.GetSpansAsync(state.Document, cancellationToken).ConfigureAwait(false);
+        return spans;
+    }
+
+    public override Task<object?> ComputeExpensiveVersionAsync(SpellCheckState state, CancellationToken cancellationToken)
+    {
+        return SpecializedTasks.Null<object>();
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/SpellCheck/SpellCheckPullCache.cs:40]: doc what this means.

File: src/LanguageServer/ProtocolUnitTests/Diagnostics/DiagnosticsPullCacheTests.cs [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+using Microsoft.CodeAnalysis.SolutionCrawler;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+using Xunit.Abstractions;
+using LSP = Roslyn.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.UnitTests.Diagnostics;
+public class DiagnosticsPullCacheTests(ITestOutputHelper testOutputHelper) : AbstractPullDiagnosticTestsBase(testOutputHelper)
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/ProtocolUnitTests/Diagnostics/DiagnosticsPullCacheTests.cs:23]:
```suggestion

public sealed class DiagnosticsPullCacheTests(ITestOutputHelper testOutputHelper)
    : AbstractPullDiagnosticTestsBase(testOutputHelper)
```

File: src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs (hunk 10 above)
Comment (dibarbet) [reply src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs:62]: Definitely potentially possible.  I'll leave that to investigate in a different PR.

File: src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs (hunk 11 above)
Comment (dibarbet) [reply src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs:57]: this is the operation progress stuff.  We always recompute until we've fully loaded the sln.

File: src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+                WriteLinePosition(span.Start, writer);
+                WriteLinePosition(span.End, writer);
+            }
+
+            static void WriteLinePosition(LinePosition position, ObjectWriter writer)
+            {
+                writer.WriteInt32(position.Line);
+                writer.WriteInt32(position.Character);
+            }
+
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs:122]:
```suggestion
```

File: src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.cs [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+
+/// <summary>
+/// Specialized cache used by the 'pull' LSP handlers.  Supports storing data to know when to tell a client
+/// that existing results can be reused, or if new results need to be computed.  Multiple keys can be used,
+/// with different computation costs to determine if the previous cached data is still valid.
+/// </summary>
+internal abstract partial class VersionedPullCache<TCheapVersion, TExpensiveVersion, TState, TComputedData>(string uniqueKey)
+    where TComputedData : notnull
+{
+
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.cs:24]:
```suggestion
```

File: src/LanguageServer/Protocol/Handler/SpellCheck/SpellCheckPullCache.cs [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SpellCheck;
+
+internal record struct SpellCheckState(ISpellCheckSpanService Service, Document Document);
+
+/// <summary>
+/// Simplified version of <see cref="VersionedPullCache{TCheapVersion, TExpensiveVersion, TState, TComputedData}"/> that only uses a 
+/// single cheap key to check results against.
+/// </summary>
+internal class SpellCheckPullCache(string uniqueKey) : VersionedPullCache<(Checksum parseOptionsChecksum, Checksum textChecksum)?, object?, SpellCheckState, SpellCheckSpan>(uniqueKey)
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/SpellCheck/SpellCheckPullCache.cs:19]:
```suggestion
internal sealed class SpellCheckPullCache(string uniqueKey) : VersionedPullCache<(Checksum parseOptionsChecksum, Checksum textChecksum)?, object?, SpellCheckState, SpellCheckSpan>(uniqueKey)
```

File: src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs [hunk 22]
```diff
@@ -116,6 +116,12 @@ public override int GetHashCode()
         // The checksum is already a hash. Just read a 4-byte value to get a well-distributed hash code.
         return (int)Data1;
     }
+
+    public int CompareTo(Checksum other)
+    {
+        var result = Data1.CompareTo(other.Data1);
+        return (result != 0) ? result : Data2.CompareTo(other.Data2);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs:123]:
```suggestion
        return result != 0 ? result : Data2.CompareTo(other.Data2);
```

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+                writer.WriteString(key);
+                writer.WriteString(diagnosticData.Properties[key]);
+            }
+
+            if (diagnosticData.ProjectId != null)
+                writer.WriteGuid(diagnosticData.ProjectId.Id);
+
+            WriteDiagnosticDataLocation(diagnosticData.DataLocation, writer);
+
+            foreach (var additionalLocation in diagnosticData.AdditionalLocations)
```
Comment (ToddGrun) [inline src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs:84]: >AdditionalLocations [](http://example.com/codeflow?start=62&length=19)

Are there usually very many items in here? If so, do they usually have the same file path?

File: src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+            {
+                WriteLinePosition(span.Start, writer);
+                WriteLinePosition(span.End, writer);
+            }
+
+            static void WriteLinePosition(LinePosition position, ObjectWriter writer)
+            {
+                writer.WriteInt32(position.Line);
+                writer.WriteInt32(position.Character);
+            }
```
Comment (ToddGrun) [inline src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs:121]: nit: trailing empty line

File: src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs (hunk 23 above)
Comment (ToddGrun) [reply src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs:84]: Oh, I didn't notice there were multiple file spans per location. Maybe these can't be easily grouped by file location.

**Pull request merged**